sources/mdb_iamproxy.c

```diff
@@ -0,0 +1,254 @@
+{
+	for (int i = 0; i < MDB_IAMPROXY_DEFAULT_HEADER_SIZE; ++i) {
+		dst[i] = (src & 0xFF);
+		src >>= CHAR_BIT;
+	}
+}
+
+void fetch_header(uint64_t *dst, char src[])
+{
+	for (int i = 0; i < MDB_IAMPROXY_DEFAULT_HEADER_SIZE; ++i) {
+		(*dst) |= (((uint64_t)src[i]) << (i * CHAR_BIT));
+	}
+}
+
+machine_msg_t *mdb_iamproxy_io_read(machine_io_t *io)
+{
+	machine_msg_t *header;
+	machine_msg_t *msg;
+
+	uint64_t body_size = 0;
+	uint64_t received = 0;
+
+	/* RECEIVE HEADER */
+	header = machine_read(io, MDB_IAMPROXY_DEFAULT_HEADER_SIZE,
+			      MDB_IAMPROXY_DEFAULT_RECEIVING_HEADER_TIMEOUT);
+	if (header == NULL) {
+		return NULL;
+	}
+	fetch_header(&body_size, (char *)machine_msg_data(header));
+	machine_msg_free(header);
+
+	if (body_size > MDB_IAMPROXY_MAX_MSG_BODY_SIZE) {
+		return NULL;
+	}
+	msg = machine_read(io, body_size,
+			   MDB_IAMPROXY_DEFAULT_RECEIVING_BODY_TIMEOUT);
+	if (msg == NULL) {
+		return NULL;
+	}
+
+	return msg;
+}
+
+int mdb_iamproxy_io_write(machine_io_t *io, machine_msg_t *msg)
+{
+	/*GET COMMON MSG INFO AND ALLOCATE BUFFER*/
+	int32_t send_result = MDB_IAMPROXY_RES_OK;
+	uint64_t body_size = machine_msg_size(
+		msg); // stores size of message (add one byte for 'c\0')
+
+	/* PREPARE HEADER BUFFER */
+	machine_msg_t *header =
+		machine_msg_create(MDB_IAMPROXY_DEFAULT_HEADER_SIZE);
+	if (header == NULL) {
+		send_result = MDB_IAMPROXY_RES_ERROR;
+		goto free_end;
+	}
+	put_header((char *)machine_msg_data(header), body_size);
+
+	/*SEND HEADER TO SOCKET*/
+	if (machine_write(io, header, MDB_IAMPROXY_DEFAULT_SENDING_TIMEOUT) <
+	    0) {
+		send_result = MDB_IAMPROXY_RES_ERROR;
+		goto free_end;
+	}
+
+	/*SEND MSG TO SOCKET*/
+	if (machine_write(io, msg, MDB_IAMPROXY_DEFAULT_SENDING_TIMEOUT) < 0) {
+		send_result = MDB_IAMPROXY_RES_ERROR;
+		goto free_end;
+	}
+
+free_end:
+	return send_result;
+}
+
+int mdb_iamproxy_authenticate_user(const char *username, const char *token,
+				   od_instance_t *instance, od_client_t *client)
+{
+	int32_t authentication_result =
+		MDB_IAMPROXY_CONN_DENIED; // stores authenticate status for user (default value: CONN_DENIED)
+	int32_t correct_sending =
+		MDB_IAMPROXY_CONN_ACCEPTED; // stores stutus of sending data to iam-auth-proxy
+	char *auth_status_char;
+	machine_msg_t *msg_username = NULL, *msg_token = NULL,
+		      *auth_status = NULL, *external_user = NULL;
+
+	/*SOCKET SETUP*/
+	struct sockaddr *saddr;
+	struct sockaddr_un
+		exchange_socket; // socket for interprocceses connection
+	memset(&exchange_socket, 0, sizeof(exchange_socket));
+	exchange_socket.sun_family = AF_UNIX;
+	saddr = (struct sockaddr *)&exchange_socket;
+	od_snprintf(exchange_socket.sun_path, sizeof(exchange_socket.sun_path),
+		    "%s", MDB_IAMPROXY_DEFAULT_SOCKET_FILE);
+
+	/*SETUP IO*/
+	machine_io_t *io;
+	io = machine_io_create();
+	if (io == NULL) {
+		authentication_result = MDB_IAMPROXY_CONN_ERROR;
+		goto free_end;
+	}
+
+	machine_set_nodelay(io, instance->config.nodelay);
```

Comment:
`no_delay` and `keep_alive` is for TCP sockets, they are not needed for UNIX sockets